cxx/structureTranslationModule/translator/NrelFromNodeTranslator.cpp

```diff
@@ -16,45 +18,93 @@ NrelFromNodeTranslator::NrelFromNodeTranslator(ScMemoryContext * context)
   ScAddr sourceNode;
   ScAddr node;
   ScAddr nrelNode;
 
   ScTemplate scTemplate;
-  scTemplate.Triple(structAddr, ScType::EdgeAccessVarPosPerm, ScType::EdgeDCommonVar >> TranslationConstants::EDGE_ALIAS);
+  scTemplate.Triple(
+      structAddr, ScType::EdgeAccessVarPosPerm, ScType::EdgeDCommonVar >> TranslationConstants::EDGE_ALIAS);
   scTemplate.Quintuple(
       ScType::NodeVar >> TranslationConstants::SOURCE_ALIAS,
       TranslationConstants::EDGE_ALIAS,
       ScType::NodeVar >> TranslationConstants::NODE_ALIAS,
       ScType::EdgeAccessVarPosPerm,
       ScType::NodeVarNoRole >> TranslationConstants::NREL_ALIAS);
-  context->HelperSmartSearchTemplate(scTemplate,
-  [&](ScTemplateResultItem const & searchResult)
-  {
-    node = searchResult[TranslationConstants::NODE_ALIAS];
-    sourceNode = searchResult[TranslationConstants::SOURCE_ALIAS];
-    nrelNode = searchResult[TranslationConstants::NREL_ALIAS];
-    std::string const & nodeMainIdtf = utils::CommonUtils::getMainIdtf(context, node, {TranslationKeynodes::lang_en});
-    if (nodeMainIdtf.empty())
-      return ScTemplateSearchRequest::CONTINUE;
-    std::string const & sourceMainIdtf = utils::CommonUtils::getMainIdtf(context, sourceNode, {TranslationKeynodes::lang_en});
-    if (sourceMainIdtf.empty())
-      return ScTemplateSearchRequest::CONTINUE;
-    std::string const & nrelMainIdtf = utils::CommonUtils::getMainIdtf(context, nrelNode, {TranslationKeynodes::lang_en});
-    if (nrelMainIdtf.empty())
-      return ScTemplateSearchRequest::CONTINUE;
-    translations << sourceMainIdtf << " "  << nrelMainIdtf << " "  << nodeMainIdtf << ". ";
-    return ScTemplateSearchRequest::CONTINUE;
-  },
-  [&](ScAddr const & element)
-  {
-    return isInStructure(structAddr, element);
-  });
-
+  context->HelperSmartSearchTemplate(
+      scTemplate,
+      [&](ScTemplateResultItem const & searchResult)
+      {
+        node = searchResult[TranslationConstants::NODE_ALIAS];
+        sourceNode = searchResult[TranslationConstants::SOURCE_ALIAS];
+        nrelNode = searchResult[TranslationConstants::NREL_ALIAS];
+        if (context->HelperCheckEdge(
+                TranslationKeynodes::translation_ignored_keynodes, nrelNode, ScType::EdgeAccessConstPosPerm))
+          return ScTemplateSearchRequest::CONTINUE;
+        std::string const & nrelMainIdtf =
+            utils::CommonUtils::getMainIdtf(context, nrelNode, {scAgentsCommon::CoreKeynodes::lang_ru});
+        if (nrelMainIdtf.empty())
+          return ScTemplateSearchRequest::CONTINUE;
+        std::string const & sourceMainIdtf =
+            utils::CommonUtils::getMainIdtf(context, sourceNode, {scAgentsCommon::CoreKeynodes::lang_ru});
+        if (sourceMainIdtf.empty())
+          return ScTemplateSearchRequest::CONTINUE;
+        if (used.find(nrelMainIdtf) == used.end())
+        {
+          used.insert(nrelMainIdtf);
+          relationTranslation = get_translation_of_relation(structAddr, nrelNode, sourceNode);
```

Comment:
I didn't get it, why do you need this if and get_translation_of_relation method?

Reply:
There are relations with many targets from one source, like

rocks
=> nrel_strict_inclusion:
    sedimentary_rocks;
    igneous_rocks;
    metamorphic_rocks;;

The point of this change is that once such relation is found, all target nodes with this relations are searched and listed in translation with commas. This if is for not repeating the same information. And all this helps not to mix up nodes of different such relations and make translation look good.